### ISSUE TITLE
User Interface - Extend "sort by" in campaign listing

### ISSUE THREAD

**Issue by shevegen** [wesnoth/wesnoth#3183]
Hello.

Right now we can sort campaigns in the main menu interface via:

a) Name
b) Dates

However had, I would like to have the possibility to sort by:

**Amount of scenarios**

The reason as to why is simple - I would rather avoid campaigns
that are too short, like 4-9 scenarios. I much prefer campaigns
with +10 scenarios sinec I can play them in one-go.

Would it be possible to add it? If there is not enough space, 
perhaps a drop-down (arrow down) button could appear to the
right to allow for additional sorting algorithms to be used).

Hope I could explain what I meant there. The sort entries are
at "Play a Campaign".

**Comment by stevecotton** [comment]
I think sorting based on duration makes sense, but I think number of scenarios isn't a reliable measure. Take for example PR #3125 which is that the current final scenario is more than half of the entire campaign's play-time.

Having said that, I don't know how it should be measured, unless reviewers start timing themselves and then comparing to a standard unit (say, the amount of time it takes to play through HttT).

**Comment by CelticMinstrel** [comment]
This is a little tricky, since a campaign's length may not be a fixed number - it can depend on your choices during the campaign. In most cases it's probably possible to procedurally determine a lower and upper bound on the number of scenarios (by looking through all events defined in each scenario, including nested ones, and parsing the `[endlevel]` tags), but this can't possibly work in every imagineable case. Even in mainline, Descent into Darkness is a problematic case (as the last scenario is set to be played through multiple times before ending), and any campaign which uses `[insert_tag]` to set up and `[endlevel]` would also break such an algorithm.

So basically there are three possibilities:

1. Use an algorithm to try and guess at the length, attempting to break cycles somehow and accepting that unusual use cases will just give inaccurate results.
2. Just count the total number of scenarios in the campaign. This would overestimate the length for any campaign with branching paths.
3. Let the campaign designer tell us how many scenarios there are. This means they could lie about the length.

**Comment by Wedge009** [comment]
I didn't see it already mentioned that there are also purely story-only scenarios in some campaigns. I don't think these are included in the scenario counts reported in a campaign's description, but would be another thing to consider if trying to determine number of scenarios programatically.
